### PR TITLE
image: write resolved-content from snap prepare-image

### PR DESCRIPTION
**Pull request by mvo5**
This commit adds code to create a new "resolved-content" directory
under opts.PrepareDir that can be picked up by ubuntu-image. This
dir will contain the resolved content for each volume structure
that should be written to the image.

The corresponding ubuntu-image code would be something like https://github.com/CanonicalLtd/ubuntu-image/compare/master...mvo5:snap-prepare-image-resolved-content?expand=1

### CODE REVIEW

File: image/helpers.go [hunk 1]
```diff
@@ -404,3 +405,25 @@ func (tsto *ToolingStore) Find(at *asserts.AssertionType, headers map[string]str
...
+// var so that it can be mocked for tests
+var writeResolvedContent = writeResolvedContentImpl
+
+// XXX: move to gadget?
+func writeResolvedContentImpl(targetdir, gadgetUnpackDir, kernelUnpackDir string, model *asserts.Model) error {
+	lv, err := gadget.LaidOutSystemVolumeFromGadget(gadgetUnpackDir, kernelUnpackDir, model)
+	if err != nil {
+		return err
+	}
+	for _, ps := range lv.LaidOutStructure {
```
Comment (pedronis) [inline image/helpers.go:418]: for UC20 this code will write files that are not needed, because we really need to take care only of system-seed, no? OTOH maybe it's ok because is ready if we do some of install work already at ubuntu-image time later.

ATM the layouts are already different:

UC16/18
```
<prepare-dir>/image
```

UC20
```
<prepare-dir>/system-seed
```

File: image/helpers.go [hunk 2]
```diff
@@ -404,3 +406,56 @@ func (tsto *ToolingStore) Find(at *asserts.AssertionType, headers map[string]str
...
+			}
+			dst := filepath.Join(targetdir, volName, ps.Name)
+			// on UC20, ensure system-seed links back to the
+			// <PrepareDir>/system-seed
+			if ps.Role == gadget.SystemSeed {
+				uc20systemSeedDir := filepath.Join(prepareDir, "system-seed")
+				if err := os.MkdirAll(filepath.Dir(dst), 0755); err != nil {
+					return err
+				}
+				if err := os.Symlink(uc20systemSeedDir, dst); err != nil {
```
[the remark this answers is not in the document]
Comment (anonymouse64) [reply image/helpers.go:454]: though to be fair this was with current edge channel of ubuntu-image, the error is as follows:

```
error: cannot write filesystem content of source:boot-assets/: cannot create prefix directory: mkdir ubuntu-image-work-dir/unpack/resolved-content/pi/ubuntu-seed: file exists
```
and the symlink looks like this:

```
ubuntu-image-work-dir/unpack/resolved-content/pi$ ls -lah
total 8.0K
drwxr-xr-x 2 ijohnson ijohnson 4.0K Feb 25 17:08 .
drwxr-xr-x 3 ijohnson ijohnson 4.0K Feb 25 17:08 ..
lrwxrwxrwx 1 ijohnson ijohnson   40 Feb 25 17:08 ubuntu-seed -> ubuntu-image-work-dir/unpack/system-seed
```

File: image/helpers.go [hunk 3]
```diff
@@ -404,3 +406,56 @@ func (tsto *ToolingStore) Find(at *asserts.AssertionType, headers map[string]str
...
 	return tsto.sto.Assertion(at, pk, tsto.user)
 }
+
+// var so that it can be mocked for tests
+var writeResolvedContent = writeResolvedContentImpl
+
+// XXX: move to gadget?
+func writeResolvedContentImpl(prepareDir, gadgetUnpackDir, kernelUnpackDir string, model *asserts.Model) error {
+	// XXX: ugly this is hardcoded here and that the caller needs
+	// to know what we did here :/
```
Comment (anonymouse64) [inline image/helpers.go:416]: this sentence is hard to parse, did you mean to say something like

```suggestion
	// XXX: this is ugly; we hardcode this path here because the caller needs
	// to know what we did here :/
```

File: image/helpers_test.go [hunk 4]
```diff
@@ -130,3 +132,93 @@ func (s *imageSuite) TestDownloadSnap(c *check.C) {
...
+		{"meta/gadget.yaml", validGadgetYaml},
+		{"system-seed.efi", "content of system-seed.efi"},
+		{"grubx64.efi", "content of grubx64.efi"},
+		{"foo", "content of foo"},
+		{"non-fs.img", "content of non-fs.img"},
+	})
+	kernelRoot := c.MkDir()
+	model := s.makeUC20Model(nil)
+
+	err = image.WriteResolvedContent(prepareImageDir, gadgetRoot, kernelRoot, model)
```
Comment (anonymouse64) [inline image/helpers_test.go:198]: we need a unit test for prepareImageDir being a relative dir too

**Review by anonymouse64 — COMMENTED** [review]
this looks good to me, I pushed one change to fix the symlink issue I mention below, but otherwise I think it's good with an additional unit test

This also fully enables my uc20 pi-config work too so +2 from me :smile:

File: image/helpers.go (hunk 1 above)
Comment (mvo5) [reply image/helpers.go]: Yeah, this will produce too much data on uc20 currently (it. should stop after it found the system-seed).

File: image/helpers.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mvo5) [reply image/helpers.go:451]: Thanks for fixing this!

File: image/image_linux.go [hunk 5]
```diff
@@ -406,18 +417,30 @@ func setupSeed(tsto *ToolingStore, model *asserts.Model, opts *Options) error {
...
+		return err
+	}
 
 	if err := boot.MakeBootable(model, bootRootDir, bootWith, nil); err != nil {
 		return err
 	}
 
+	// write resolved content to structure root
+	// XXX: move to gadget?
+	if !opts.Classic {
```
[the remark this answers is not in the document]
Comment (mvo5) [reply image/image_linux.go:438]: Good catch, thank you!

File: image/helpers.go (hunk 3 above)
Comment (mvo5) [reply image/helpers.go]: I removed this now and tweaked the signature of the function

File: image/helpers_test.go (hunk 4 above)
Comment (mvo5) [reply image/helpers_test.go]: I added that now

File: image/helpers.go (hunk 1 above)
Comment (mvo5) [reply image/helpers.go]: It maybe okay for now that prepare-image just outputs all the resolved content and ubuntu-image picks what it needs. Longer term ubuntu-image will be in go and it will call better and richer APIs than a (very limited) CLI itnerface.

File: image/helpers.go [hunk 6]
```diff
@@ -404,3 +406,57 @@ func (tsto *ToolingStore) Find(at *asserts.AssertionType, headers map[string]str
...
+var writeResolvedContent = writeResolvedContentImpl
+
+// XXX: move to gadget?
+func writeResolvedContentImpl(targetDir, prepareDir, gadgetUnpackDir, kernelUnpackDir string, model *asserts.Model) error {
+	fullPrepareDir, err := filepath.Abs(prepareDir)
+	if err != nil {
+		return err
+	}
+
+	info, err := gadget.ReadInfoAndValidate(gadgetUnpackDir, model, nil)
```
Comment (pedronis) [inline image/helpers.go:420]: there's an argument to be made that this function should receive the gadget.Info directly instead of the model.

the arguments would be:

targetDir, info, gadgetUnpackDir, kernelUnpackDir

**Review by pedronis — COMMENTED** [review]
looks good, some comments mostly around the signature of writeResolvedContent

File: image/image_linux.go [hunk 7]
```diff
@@ -130,6 +130,13 @@ func unpackGadget(gadgetFname, gadgetUnpackDir string) error {
 	return snap.Unpack("*", gadgetUnpackDir)
 }
 
+func unpackKernel(kernelFname, kernelUnpackDir string) error {
```
[the remark this answers is not in the document]
Comment (mvo5) [reply image/image_linux.go:133]: Indeed, silly me. Removed the duplicated helper.

File: image/helpers.go (hunk 6 above)
Comment (mvo5) [reply image/helpers.go:420]: Indeed, I like that. I updated the code and had to tweak some tests. Note that the signature is slightly different (not targetDir) as I applied the earlier feedback about just using gadgetUnpackDir. Let me know if you feel it should be done differently.

**Review by pedronis — APPROVED** [review]
thank you

**Review by anonymouse64 — APPROVED** [review]
lgtm, thanks for the additional unit tests here

**Comment by anonymouse64** [comment]
(I pushed a fix for the failing unit test which was racy I think since the output wasn't sorted)

File: image/helpers.go [hunk 8]
```diff
@@ -404,3 +406,57 @@ func (tsto *ToolingStore) Find(at *asserts.AssertionType, headers map[string]str
...
+// /tmp/prep-img/resolved-content/pi/ubuntu-seed/{config.txt,bootcode.bin,...}
+func writeResolvedContentImpl(prepareDir string, info *gadget.Info, gadgetUnpackDir, kernelUnpackDir string) error {
+	fullPrepareDir, err := filepath.Abs(prepareDir)
+	if err != nil {
+		return err
+	}
+	targetDir := filepath.Join(fullPrepareDir, "resolved-content")
+
+	constraints := gadget.LayoutConstraints{
+		NonMBRStartOffset: 1 * quantity.OffsetMiB,
```
Comment (bboozzoo) [inline image/helpers.go:428]: It's pretty annoying that we need to keep repeating this across the codebase. Maybe we should add something like
```
var DefaultLayoutConstraints = LayoutConstraints{
	NonMBRStartOffset: 1 * quantity.OffsetMiB,
}
```
in the gadget package. I'll take a look into this in a separate PR.

**Review by bboozzoo — APPROVED** [review]
LGTM

File: image/helpers.go (hunk 8 above)
Comment (mvo5) [reply image/helpers.go:428]: Great suggestions, I will also do this via a followup

**Comment by mvo5** [comment]
After spending some quality time with ubuntu-image it looks like this needs a little bit of tweaking because volume structures often do not have names and ubuntu-image uses the structure number instead. So prepare-image should layout as "<volume-name>/<structure-number>" by default

File: image/helpers.go (hunk 8 above)
Comment (mvo5) [reply image/helpers.go]: Actually I did it in this one now.

**Comment by mvo5** [comment]
Fwiw, I tested this with the coresponding ubuntu-image changes in https://github.com/CanonicalLtd/ubuntu-image/compare/master...mvo5:snap-prepare-image-resolved-content?expand=1 and got a working image that had the DTBs, seeded and subsequent kernel/gadget updates cause no observable issues.

**Comment by mvo5** [comment]
@pedronis I did a small tweak to the dir structure in my latest commit, it uses `resolved-content/<vol-name>/part<NR>/` now (e.g. resolved-content/pi/part1) because a lot of the volume structures do not have names and ubuntu-image is already doing the "part<NR>" bit internal. Would love to get your okay on this particular chagne before I merge this PR.

File: image/helpers.go [hunk 9]
```diff
@@ -412,10 +412,10 @@ var writeResolvedContent = writeResolvedContentImpl
 // writeResolvedContent takes gadget.Info and the unpacked
 // gadget/kernel snaps and outputs the resolved content from the
 // {gadget,kernel}.yaml into a filesystem tree with the structure:
-// <prepareImageDir>/resolved-content/<volume-name>/<structure-name>/...
+// <prepareImageDir>/resolved-content/<volume-name>/<structure-nr>/...
```
Comment (pedronis) [inline image/helpers.go:415]: this is missing "part" ?

File: image/helpers.go [hunk 10]
```diff
@@ -412,10 +412,10 @@ var writeResolvedContent = writeResolvedContentImpl
...
 //
 // E.g.
-// /tmp/prep-img/resolved-content/pi/ubuntu-seed/{config.txt,bootcode.bin,...}
+// /tmp/prep-img/resolved-content/pi/part1/{config.txt,bootcode.bin,...}
```
Comment (pedronis) [inline image/helpers.go:418]: is ubuntu-seed the first structure?  then it should be part0 ?

**Review by pedronis — APPROVED** [review]
looks ok as we discussed, some nitpicks about the comment

**Review by anonymouse64 — COMMENTED** [review]
post merge +1 from me again
